File: libraries/Microsoft.CommonExpressions/Expressions/ExpressionEvaluator.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

namespace Microsoft.Expressions
{
    public delegate void ValidateExpressionDelegate(Expression expression);
    public delegate (object value, string error) EvaluateExpressionDelegate(Expression expression, object state);

    public class ExpressionEvaluator
    {
        public ExpressionEvaluator(EvaluateExpressionDelegate evaluator,
            ExpressionReturnType returnType = ExpressionReturnType.Object,
            ValidateExpressionDelegate validator = null)
        {
            _evaluator = evaluator;
            ReturnType = returnType;
            _validator = validator ?? new ValidateExpressionDelegate((expr) => { });
        }

        private ValidateExpressionDelegate _validator;
        private EvaluateExpressionDelegate _evaluator;

        public (object value, string error) TryEvaluate(Expression expression, object state)
            => _evaluator(expression, state);

        public void ValidateExpression(Expression expression)
            => _validator(expression);

        public ExpressionReturnType ReturnType { get; }
    }
}

File: libraries/AdaptiveExpressionsCpp/AdaptiveExpressionsCpp/Parser/ExpressionEvaluator.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

#include "ExpressionEvaluator.h"
#include "../Parser/FunctionUtils.h"
#include "../BuiltinFunctions/Subtract.h"

ExpressionEvaluator::ExpressionEvaluator(std::string type, ReturnType returnType)
{
    m_type = type;
    m_returnType = returnType;
}

void ExpressionEvaluator::setReturnType(ReturnType returnType)
{
    m_returnType = returnType;
}

ReturnType ExpressionEvaluator::getReturnType()
{
    return m_returnType;
}

